### PR TITLE
[LS] Improve loading package listener metadata using a ls-package index

### LINKED ISSUE #43172: [Improvement]: Reduce the LS Package Loader Central Packages Loading time
### Description

Currently, the Language Server  package loader is responsible for loading packages under the ballerinax organization using the Central API client. This approach loads information about only 10 packages at a time, which significantly increases the time taken to complete the LS package loading process, ultimately resulting in a poor developer experience (DX).

### Describe your problem(s)

_No response_

### Describe your solution(s)

To improve this DX and reduce the central package loading time, we have identified three potential approaches:

- **Load All Details from a Remote JSON File**
In this approach, package information is pre-encoded into a JSON file hosted in a remote GitHub repository. This JSON file is updated daily through a scheduled GitHub Action. The Language Server will fetch the remote JSON containing package information and use it to populate the ModuleInfo objects.

Advantages:
Faster than fetching packages individually since the information is pre-computed.
The JSON file size can be optimized by removing unnecessary attributes, reducing the data transfer load.

Disadvantages:
Users may not receive the most up-to-date information about available packages, as the data is only updated once daily.

- **Load All Details Through a GraphQL API Provided by Ballerina Central**
In this approach, package information is retrieved from the Ballerina Central GraphQL API. This API can provide all package information using a single GraphQL query, and the LS can use the response to populate the ModuleInfo objects.

Advantages:
Easy to maintain as it leverages an existing API.
Ensures users receive up-to-date package information.

Disadvantages:
Relatively slower compared to the first approach due to the performance of the central API.

- **Increase the Limit of the Current Request**
Currently, the loader retrieves 10 packages at a time, resulting in a loading time of over 60 seconds for all ballerinax packages. [Increasing this limit to 1000](https://github.com/ballerina-platform/ballerina-lang/blame/de5060eea0cbefe3c7fe76fcafda5ddecc8c8d8a/language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/CentralPackageDescriptorLoader.java#L89) packages would reduce the loading time to less than 30 seconds.

Advantages:

Reduces the overall loading time significantly.
Disadvantages:
Relatively slower compared to the first approach due to the performance of the central API.

Latency comparison between REST and GraphQL apis haven't been done. Since we get more precise and compressed information though the GraphQL API it should be faster.

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #43172

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java [hunk 1]
```diff
@@ -154,21 +158,78 @@ public List<LSCompletionItem> getServiceTemplates(BallerinaCompletionContext ctx
...
+//
+//        SnippetContext snippetContext = new SnippetContext(serviceSnippet.currentSnippetIndex - 1);
+//
+//        if (!serviceSnippet.unimplementedMethods.isEmpty()) {
+//            for (MethodSymbol methodSymbol : serviceSnippet.unimplementedMethods) {
+//                String functionSnippet =
+//                        generateMethodSnippet(importsAcceptor, methodSymbol, snippetContext, context);
+//                methodSnippets.add(functionSnippet);
+//            }
+//        }
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java:206]: Do we need this?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java [hunk 2]
```diff
@@ -154,21 +158,78 @@ public List<LSCompletionItem> getServiceTemplates(BallerinaCompletionContext ctx
...
         });
         return completionItems;
     }
 
+    private LSCompletionItem generateServiceSnippetIndexedListener(IndexedListenerMetaData indexedListenerMetaData,
+                                                    BallerinaCompletionContext context) {
+
+        String symbolReference;
+        ImportsAcceptor importsAcceptor = new ImportsAcceptor(context);
+        PackageID packageID = new PackageID(new Name(indexedListenerMetaData.lsModuleId().orgName()),
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java:178]: Extract `indexedListenerMetaData.lsModuleId()` into a separate variable.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSPackageLoader.java [hunk 3]
```diff
@@ -350,35 +360,179 @@ public List<ModuleInfo> updatePackageMap(DocumentServiceContext context) {
...
+            String userHomeIndexFileChecksum = getFileChecksum(BALLERINA_USER_HOME_INDEX.toString());
+            String payload = "";
+            boolean indexUpdated = false;
+            try {
+                String expectedCheckSum = this.centralPackageDescriptorLoader.getLSPackageIndexChecksum();
+                if (!userHomeIndexFileChecksum.equals(expectedCheckSum)) {
+                    // Download the file from the central and load the listeners
+                    payload = this.centralPackageDescriptorLoader.getLSPackageIndex();
+                    lsListenerIndex = new Gson().fromJson(payload, LSListenerIndex.class);
+                    indexUpdated = true;
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSPackageLoader.java:418]: We can call `saveLSPackageIndexInBallerinaUserHome` here and we don't need `indexUpdated`.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java (hunk 1 above)
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java:206]: no, we don't.

**Pull request closed without merging**